activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -97,7 +97,16 @@ def require_counter_update?
         end
 
         def replace_keys(record)
-          owner[reflection.foreign_key] = record ? record._read_attribute(primary_key(record.class)) : nil
+          owner[reflection.foreign_key] = record ? record._read_attribute(inverse_primary_key(record) || primary_key(record.class)) : nil
+        end
+
+        def inverse_primary_key(record)
+          inverse = inverse_reflection_for(record) || reverse_inverse_reflection_for(record)
+          inverse && inverse.options[:primary_key]
+        end
+
+        def reverse_inverse_reflection_for(record)
+          record._reflections.values.find { |record_reflection| record_reflection.options[:inverse_of] == reflection.name }
```

Comment:
I fell this is doing too much. If the `inverse_reflection_for(record)` to find the primary key we should just raise and ask the developer to set the right primary key in the inversion.

Reply:
Do you mean definition like below is not allowed after rails 5.2?
cc @LeandroSNunes

```rb
class Product < ActiveRecord::Base
  self.primary_key = :id

  has_many :pictures, as: :imageable, primary_key: :code, inverse_of: :imageable
end
```

Reply:
I think this one we can know because you are defining the `primary_key` in the `has_many` association. What I don't understand is what this `reverse_inverse_reflection_for` is looking for in that example. Is it trying to get the `product` association on `pictures`?

Reply:
>  Is it trying to get the product association on pictures?
Yes, it is.

Reply:
I see. This is doing too much and it is really hard to understand what is the expected behavior by the tests on this PR and also in the original issue. I feel there is an easier fix by either asking more options to be set in the association or in a different layer. If you look the commit that you said it broke this behavior you will see that the code before was not doing any of this extra work.